### PR TITLE
Abstract validation strategy

### PR DESCRIPTION
**Pull request by marcinwyszynski**
## Description of the change

This change introduces a new request validation strategy - a blocklist, which is less sensitive to backward incompatibilities than the explicit allowlist. This is the new default, and the blocklist can be loaded from a YAML file. The explicit allowlist is still an option, but requires explicitly passing a boolean flag.

## Type of change

- [x] New feature (non-breaking change that adds functionality);

### Development

- [x] Lint rules pass locally;
- [x] All tests related to the changed code pass in development;

### Code review

- [x] This pull request has a descriptive title and sufficient context for a reviewer. There may be a screenshot or screencast attached;
- [x] This pull request is no longer a draft;

### Deployment

- [x] Selected merge strategy is [squash merge](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-pull-request-commits);
- [ ] Changes have been reviewed by at least one other engineer;

### CODE REVIEW

File: privatevcs/validation/blocklist/rule.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	Name string `json:"name"`
+
+	// The HTTP method to match.
+	Method string `json:"method"`
+
+	// The regular expression to match the path against.
+	Path string `json:"path"`
+
+	// The regular expression to match the vendor against.
+	Vendor string `json:"vendor"`
```
Comment (adamconnelly) [inline privatevcs/validation/blocklist/rule.go:23]: I think adding vendor at the rule level is maybe overcomplicating things to begin with - at the moment we need to specify the vendor globally for a single instance of the agent, so it wouldn't make sense to set the vendor to GitHub and allow a mixture of GitHub and GitLab rules for example.

File: agent/agent.go [hunk 2]
```diff
@@ -123,24 +119,8 @@ func (a *Agent) handleRequest(ctx *spcontext.Context, id string, msg *privatevcs
...
-		"raw_path", req.URL.EscapedPath(),
-		"path", req.URL.Path,
-	)
-
-	for key, value := range msg.Headers {
-		req.Header.Set(key, value)
-	}
-
-	name, project, subdomain, err := validation.MatchRequest(a.vendor, req)
+	ctx, err = a.validator.Validate(ctx, a.vendor, req)
```
Comment (adamconnelly) [inline agent/agent.go:122]: I think we might want to still add `method`, `path`, etc to the context in here rather than relying on the validator to do so. The reason is that we log successful as well as failed requests, and want to make sure we're logging all the useful info.

**Review by adamconnelly — COMMENTED** [review]
I know we'd need to sort tweak the naming slightly, but maybe we should allow users to specify whether it's an allow list or block list? That way it's still possible to completely lock things down if that's important to you.

File: agent/agent.go [hunk 3]
```diff
@@ -123,24 +119,8 @@ func (a *Agent) handleRequest(ctx *spcontext.Context, id string, msg *privatevcs
 		}
 	}
 
-	ctx = ctx.With(
```
Comment (cube2222) [inline agent/agent.go:126]: I think this with shouldn't be removed, as it's not replicated in the new matcher (nor should it be).

File: privatevcs/validation/blocklist/rule.go (hunk 1 above)
Comment (marcinwyszynski) [reply privatevcs/validation/blocklist/rule.go:23]: Fair.

File: agent/agent.go (hunk 3 above)
Comment (marcinwyszynski) [reply agent/agent.go:126]: Correct, I removed a bit too much.

File: agent/agent.go (hunk 2 above)
Comment (marcinwyszynski) [reply agent/agent.go:122]: Correct, I removed a bit too much.

**Comment by marcinwyszynski** [comment]
@adamconnelly this is still possible using flags.

**Comment by adamconnelly** [comment]
@marcinwyszynski RE

> this is still possible using flags.

What I meant was instead of just being able to use the existing static allow list, or the dynamic block list, allowing your new dynamic strategy to be used to either build an allow list or a block list. But Kuba explained on Slack that you guys had already discussed this and come to the conclusion it wasn't necessary, at least initially.

**Comment by marcinwyszynski** [comment]
Ah OK, gotcha @adamconnelly.
The new interface makes it easy to add this as an option in the future, possibly replacing the previous approach with an explicit allowlist, which could be remote (eg. uploaded to a central location).

File: privatevcs/validation/allowlist/list.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+// Validate validates the request and returns an error if the request should be
+// blocked.
+func (l List) Validate(ctx *spcontext.Context, vendor validation.Vendor, req *http.Request) (*spcontext.Context, error) {
+	name, project, subdomain, err := MatchRequest(vendor, req)
+	if err != nil {
+		return ctx.With("match_error", err), ctx.RawError(err, "invalid request")
+	}
+
+	if subdomain != nil {
```
Comment (adamconnelly) [inline privatevcs/validation/allowlist/list.go:39]: This bit is a complete pain unfortunately. What happens here is that the VCS agent has the domain name for the VCS hard-coded, so we always just apply the path to the hostname. But if you have GitHub Enterprise subdomain isolation enabled, we actually need to make a request to `codeload.<my-github-domain>` rather than just `<my-github-domain>`. Here's the commit where I added this change: https://github.com/spacelift-io/vcs-agent/commit/ccf62feb2f2bfc1f3954261ba09b6be46f0ca623

TL;DR; I think we might need to pull this code out of the validation rules and move it into agent.go, and just always do this check before making a request (i.e. if the vendor is GitHub and we're trying to download the tarball, modify the hostname if necessary). Otherwise source downloads will break for GitHub Enterprise when not using the allowlist.

**Review by adamconnelly — COMMENTED** [review]
Overall looks good, but I think we need to tweak where the handling for the GHE subdomain is happening.

**Review by cube2222 — DISMISSED** [review]
Makes sense, other than Adam's comment.

File: privatevcs/validation/allowlist/list.go (hunk 4 above)
Comment (marcinwyszynski) [reply privatevcs/validation/allowlist/list.go]: Done.

**Review by adamconnelly — APPROVED** [review]
